src/analysis/calculus/conformal.lean

```diff
@@ -0,0 +1,183 @@
+  (hf : conformal_at f x) (h : ∀ (x : X), x ∈ u → g x = f x) :
+  conformal_at g x :=
+let ⟨f', hfderiv, hf'⟩ := hf in
+  ⟨f', hfderiv.congr_of_eventually_eq (filter.eventually_eq_of_mem (hu.mem_nhds hx) h), hf'⟩
+
+lemma comp {f : X → Y} {g : Y → Z} (x : X)
+  (hg : conformal_at g (f x)) (hf : conformal_at f x) : conformal_at (g ∘ f) x :=
+begin
+  rcases hf with ⟨f', hf₁, cf⟩,
+  rcases hg with ⟨g', hg₁, cg⟩,
+  exact ⟨g'.comp f', has_fderiv_at.comp x hg₁ hf₁, cf.comp cg⟩,
+end
+
+lemma const_smul {f : X → Y} {x : X} {c : ℝ} (hc : c ≠ 0) (hf : conformal_at f x) :
+  conformal_at (c • f) x :=
+(conformal_at_const_smul hc $ f x).comp x hf
+
+lemma conformal_factor_aux {f : E → F} {x : E} (h : conformal_at f x) :
+  ∃ (c : ℝ), 0 < c ∧ ∀ (u v : E), ⟪(fderiv ℝ f x) u, (fderiv ℝ f x) v⟫ = (c : ℝ) * ⟪u, v⟫ :=
+begin
+  let p := h.differentiable_at.has_fderiv_at,
+  rcases h with ⟨f', hf, hf'⟩,
+  rcases (is_conformal_map_iff f').mp hf' with ⟨c, hc, huv⟩,
+  exact ⟨c, hc, (hf.unique p) ▸ huv⟩,
+end
+
+/-- The conformal factor of a conformal map at some point `x`. Some authors refer to this function
+    as the characteristic function of the conformal map. -/
+def conformal_factor_at {f : E → F} {x : E} (h : conformal_at f x) : ℝ :=
+classical.some (conformal_factor_aux h)
+
+lemma conformal_factor_at_pos {f : E → F} {x : E} (h : conformal_at f x) :
+  0 < conformal_factor_at h :=
+(classical.some_spec $ conformal_factor_aux h).1
+
+lemma conformal_factor_at_inner_eq_mul_inner {f : E → F} {x : E} (h : conformal_at f x) :
+  ∀ (u v : E), ⟪(fderiv ℝ f x) u, (fderiv ℝ f x) v⟫ = (conformal_factor_at h : ℝ) * ⟪u, v⟫ :=
+(classical.some_spec $ conformal_factor_aux h).2
+
+lemma conformal_factor_at_inner_eq_mul_inner' {f : E → F} {x : E} {f' : E →L[ℝ] F}
+  (h : has_fderiv_at f f' x) (H : conformal_at f x) :
+  ∀ (u v : E), ⟪f' u, f' v⟫ = (conformal_factor_at H : ℝ) * ⟪u, v⟫ :=
+(H.differentiable_at.has_fderiv_at.unique h) ▸ (classical.some_spec $ conformal_factor_aux H).2
+
+/-- If a real differentiable map `f` is conformal at a point `x`,
+    then it preserves the angles at that point. -/
+lemma preserves_angle {f : E → F} {x : E} {f' : E →L[ℝ] F}
+  (h : has_fderiv_at f f' x) (H : conformal_at f x) (u v : E) :
+  inner_product_geometry.angle (f' u) (f' v) = inner_product_geometry.angle u v :=
+let ⟨f₁, h₁, c⟩ := H in h₁.unique h ▸ c.preserves_angle u v
+
+end conformal_at
+
+lemma conformal_at_iff_is_conformal_map_fderiv {f : X → Y} {x : X} :
+  is_conformal_map (fderiv ℝ f x) ↔ conformal_at f x :=
+begin
+  split,
+  { intros H,
+    by_cases h : differentiable_at ℝ f x,
+    { exact ⟨fderiv ℝ f x, h.has_fderiv_at, H⟩, },
+    { by_cases w : nontrivial X,
+      { exfalso, rcases w with ⟨a, b, hab⟩,
+        rw [fderiv_zero_of_not_differentiable_at h] at H,
+        have : (0 : X → Y) a = (0 : X → Y) b := rfl,
+        exact hab (H.injective this), },
+      { simp_rw [not_nontrivial_iff_subsingleton, subsingleton_iff] at w,
```

Comment:
Could you please add separate lemmas saying that if `[subsingleton X]` then for any `f :  X →ₗ[ℝ] Y`, 
- `f` is differentiable, `has_fderiv_at f 0`, etc (if these don't already exist)
- `f` satisfies `is_conformal_map`